### PR TITLE
changes to sessions_controller to support MHV sign-in

### PR DESCRIPTION
**Pull request by saneshark**
- For DS Logon and ID.me we get name, dob, ssn, etc in the saml response from the identity provider and we use this to query MVI for other attributes (if LOA3). As such, vets.gov persists the user with only the saml data.

- For MHV the saml_response does not include this information. Instead we may or may not get an ICN. If we have an ICN and the user is LOA3, then we can sideload MVI and use the values returned to persist the otherwise missing name, dob, ssn, etc we fetch from MVI when persisting user.rb. All other attributes will be delegated to MVI as usual.
  - if we don't have ICN and the user is LOA3, we should probably raise an exception of some sort, because vets.gov can't really treat this user as LOA3 at all I don't think.
  - if we have an ICN but the user is NOT LOA3 (MHV Premium), we probably shouldn't load the data from MVI at all, but in theory we could just load name, dob? and just call it LOA1?? Probably should err on side of caution and raise an exception in this scenario too.
  - if we have an ICN and user is LOA3, then this is the happy path, we can sideload MVI, use it to fill out user.rb, persist user.rb, and all the other MVI attributes are properly delegated to @mvi

### CODE REVIEW

**Comment by saneshark** [comment]
@patrickvinograd tagging you to double check my thought process on this even though its just a method stub right now.

@raquelromano-gov @NatalieEMoore said you might have some bandwidth to take a look at this. I'm going to work on it a bit and experiment some, but if you would like to take a crack at it, please feel free.

File: app/controllers/v0/sessions_controller.rb [hunk 1]
```diff
@@ -116,8 +116,20 @@ def saml_user
...
 
+    # This is used in persist_session_and_user
     def new_user_from_saml
-      @new_user_from_saml ||= User.new(saml_user.to_hash)
+      @new_user_from_saml ||= begin
+        if saml_user.mhv?
+          if saml_user.icn.present?
+            # fetch name and other attributes from MVI and use in app/models/user.rb
+          else
+            # either raise an error here and log # of times this happens or attempt to load user as loa3
```
Comment (raquelromano-gov) [inline app/controllers/v0/sessions_controller.rb]: Questions are more me understanding this than comments on the approach:
1) Is another option to log that this happened so we can track in sentry, but not raise an exception for now? Then revisit if we see this happening in production
2) What are the possible results of "attempting" to load user as LOA3:
  - is there some signal that would indicate this wasn't successful?
  - if so, then would we try to load user as LOA3 and if that fails, raise an exception?

Comment (saneshark) [reply app/controllers/v0/sessions_controller.rb]: Currently if user is not "LOA3" we don't fetch from MVI at all, so I guess I was worried about deviating too far away from that current paradigm. It definitely should not be done if the user is not "premium" or the "new LOA3." But I suppose if they are LOA3 and they don't have an ICN whereby we could query for basic fields such as name, dob, gender, ssn, I was worried what the ramifications might be seeing as some of these fields are validated on user.rb as needing to be present.

To answer your questions:
1. I think we know with certainty that this will happen for some 13% of MHV users.
2. I don't know what the ramifications are. I imagine the F/E might have built up some expectation that LOA3 implies these basic values being present. If they don't have proper null handling, things could break in unexpected ways. Specs across vets-api would also likely need to be re-worked.

I'm not sure what you mean by "signal." If the saml response has an authn_context that signifies it is an MHV sign-in user and MHV characterizes it as "premium" we would need to have an ICN to query MVI or else the experience for this LOA3 could be at minimum severely limited, or at most error prone.

I'm certainly open to trying it out and seeing how it goes on staging, but wouldn't feel comfortable pushing that approach on to production. That being said, I dont know if I'm personally on a position to say that some 13% of total MHV users that *could* be affected is reason enough to delay launching?

Comment (saneshark) [reply app/controllers/v0/sessions_controller.rb]: I think Alastair has set aside some time on the calendar tomorrow for us to discuss the evolution of identity and how tightly current_user is coupled with MVI, etc. We can probably wait and discuss this and potential ramifications tomorrow.

Comment (raquelromano-gov) [reply app/controllers/v0/sessions_controller.rb]: sorry to have missed Alastair's session! I saw that on the agenda, but have been away all week.  In any case, @patrickvinograd should review since I don't want to block it and definitely want to spend some time understanding the expectations shared across vets-api and vets-website. I'll catch up with Alastair to hear what he shared

File: lib/saml/user_attributes/mhv.rb [hunk 2]
```diff
@@ -77,10 +86,50 @@ def loa_highest
...
+        mvi&.profile&.gender
+      end
+
+      # Probably need to rescue from when ICN query returns no result returning NullMVI
+      # Logging these various scenarios would provide useful data
+      def mvi
+        @mvi ||= begin
+          if PREMIUM_LOAS.include?(account_type)
+            if mhv_icn.present?
+              # What if the ICN doesn't return a hit when querying MVI???
```
Comment (kreek) [inline lib/saml/user_attributes/mhv.rb]: @saneshark does this happen? no profile returned when querying MVI by ICN? A bit confused around the need for `NullMvi`. Is it the user that can't be persisted? I think we should revisit those user validations and possibly move them as a guard against querying MVI. Probably need more context behind everything here 😄  but it's getting hard to follow with a specialized null class and some things coming from SAML and methods hitting MVI which mirroring methods in the user class `mvi&.profile&.gender` etc.

Comment (saneshark) [reply lib/saml/user_attributes/mhv.rb]: @kreek so the problem is that for MHV sign-in we don't have first_name, last_name, dob, gender, ssn, etc. So to get those values I need to side-load MVI so that I can hydrate current_user with the required attributes for LOA3, see validations on `app/models/user.rb.` 

It's true, we talked about the fact that maybe those validations could just go away, but I'm not entirely sure what the ramifications are, and furthermore, I'm concerned with how the existing exception handling on mvi in current user bubbles up to the user. By fetching MVI pre-emptively in the saml_callback I can ensure that the user is not persisted, but doing so in current_user itself would mean that the user is already persisted. And yes, an ICN value could possibly not be provided, additionally, the ICN that is provided might not resolve to an actual MVI record if what MHV has is outdated. Finally, there is no guarantee that MVI service is available, it could be down. In each of these three scenarios we need errors that are unique to the MHV sign-in experience so that we can communicate remediation to the user effectively, otherwise we're left with a very generic, you cannot login (even though your credentials are good and you're a premium MHV user). The NullMVI object is something I'm more or less just playing with here to see if it is a better pattern. It was the first solution to the problem that occurred to me. But I think it helps illustrate the problem at hand.

To be clear, I'm really not happy with this proposed implementation, but it's the least heavy handed approach I could think of until we can address authorization and current user hydration. The way mvi is delegated on current_user is a challenge to deal with currently and I didn't think I would have enough time to refactor it.

Comment (saneshark) [reply lib/saml/user_attributes/mhv.rb]: And yes, you're right. I have not yet addressed the edge cases or introduced the granular error handling I'm describing for the 3 scenarios in which a user is logged in but necessary attributes are unavailable. I was going to address those in a separate PR.

Comment (saneshark) [reply lib/saml/user_attributes/mhv.rb]: Another approach, one that I personally would advocate for, but haven't had time to explore... would be for MVI to ALWAYS be loaded in the saml callback when the user is "loa3" or "premium." and that current_user no longer needs to delegate like it does today, but it still doesn't address one important problem in that you could be signed in as an identity proofed user and not have ssn, dob, etc (should you still be characterized as LOA3 seeing as certain services authorize with the expectation of these attributes being available?)

Comment (saneshark) [reply lib/saml/user_attributes/mhv.rb]: Maybe this will be helpful, maybe not... but could be a starting point for communicating how to simplify things.

screen shot 2017-10-17 at 12 35 21 am

Comment (kreek) [reply lib/saml/user_attributes/mhv.rb]: @saneshark Thanks for that definitely sheds light on the complexity behind it. Now just need to design it so that complexity is hidden from the classes that work with authentication and authorization.

Stepping back a bit there’s really only three things going on here:
1. a user logs in from somewhere (authentication)
2. their VA profile is retrieved
3. their permissions are checked (authorization)

2 and 3 right now are a bit too intertwined in our current codebase. If we can have thwo distinct classes that act as an API/facade for profile and permissions it could simplyfy things. Then you can pipe things together Login | Profile | Permissions. 

Login, as painful as it was, is pretty much solved. A user logs in and they are in some state. That user then gets passed as input to VA Profile class to be filled out either by querying by full attributes or icn. The user that come out of that ‘pipe’ is now in another state. They want to interact with a service. The user and and the service are passed into the Permissions class which checks a policy and get a yay or nay.

MHV (and sometimes DSLogon?) seem to be here to throw a wrench into the above being simple 😀  For ID.me users MVI is not authentication, it’s just data (that’s used for authorization), but for the other two it’s authorization and authentication. Ideally MVI would never be used for authentication, despite what we were intially told it is not always the source of truth, it has duplicate records, it can be out of date, MHV ids might not line up. 

Can we make MVI irrelevant to authentication? A user still goes through the Login and Profile pipes and comes out, if they couldn’t hit MVI then the Permissions pipe always rejects them. Is that ok?

Comment (patrickvinograd) [reply lib/saml/user_attributes/mhv.rb]: The pipeline described by @kreek makes a lot of sense. 

To complicate matters I think there are some site features for which if MVI is not reachable we still want to allow certain actions. But I agree that authentication and profile are inputs (sometimes nil inputs!) to the authorization decision, and the authorization is where the complex business rules need to live.

Comment (saneshark) [reply lib/saml/user_attributes/mhv.rb]: @kreek I'm going to try to make MVI not required as part of authentication. For a non-premium user it's absolutely fine I think because the navbar on the front-end could just show the ID.me wallet email address instead of first name.  The problem arose for a premium MHV user, without first name, we wouldn't have name and other important attributes until we query MVI. The intermediary abstraction does make a lot of sense in helping to support this, but of course it touches upon a bit more refactoring than I felt comfortable making at the time.
